### PR TITLE
Allow user to use custom values on some validations tests

### PR DESCRIPTION
**Pull request by fchabouis**
Add the possibility to provide a path to a YAML file containing custom values for some tests.
This is a first step: allow to customize speeds for Excessive speed tests.
This feature was requested by the [ART](https://www.autorite-transports.fr/) who would like to run some tests with customized speed values.

usage : `cargo run -- --input gtfs.zip --custom-rules custom_rules.yml`

PS : I have done this PR following my instinct, do not hesitate to let me know if an other way was preferable!

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -292,6 +292,35 @@ You can then ask for a validation:
 * `--input` or `-i`: Path (can be a directory or a zip file) or HTTP URL (file will be downloaded) of the GTFS file.
 * `--max-issues` or `-m`: The maxium number of issues per type. Defaults to 1000.
 * `--output-format` or `-f`: Output format (when using the validator in command line). Value by default is `json`, but `yaml` is also available.
+* `--custom_rules` or `-c`: Path to a YAML file containing custom values to use during the validation.
```
Comment (thbar) [inline readme.md]: I would pick `--custom-rules` instead of `--custom_rules` for consistency!

File: readme.md [hunk 2]
```diff
@@ -292,6 +292,35 @@ You can then ask for a validation:
...
+
+## Custom rules
+Some values used during the validations can be customized by using the `--custom_rules` option and providing a path to a YAML file.
```
Comment (thbar) [inline readme.md]:
```suggestion
Some values used during the validations can be customized by using the `--custom-rules` option and providing a path to a YAML file.
```

File: readme.md (hunk 1 above)
Comment (thbar) [reply readme.md]:
```suggestion
* `--custom-rules` or `-c`: Path to a YAML file containing custom values to use during the validation.
```

File: readme.md [hunk 3]
```diff
@@ -292,6 +292,35 @@ You can then ask for a validation:
...
+| rail_speed | Maximum speed in km/h on a route_type `rail` before triggering an `ExcessiveSpeed` warning |
+| bus_speed | Maximum speed in km/h on a route_type `bus` before triggering an `ExcessiveSpeed` warning |
+| ferry_speed | Maximum speed in km/h on a route_type `ferry` before triggering an `ExcessiveSpeed` warning |
+| cable_car_speed | Maximum speed in km/h on a route_type `cable car` before triggering an `ExcessiveSpeed` warning |
+| gondola_speed | Maximum speed in km/h on a route_type `gondola` before triggering an `ExcessiveSpeed` warning |
+| funicular_speed | Maximum speed in km/h on a route_type `funicular` before triggering an `ExcessiveSpeed` warning |
+| coach_speed | Maximum speed in km/h on a route_type `coach` before triggering an `ExcessiveSpeed` warning |
+| air_speed | Maximum speed in km/h on a route_type `air` before triggering an `ExcessiveSpeed` warning |
+| taxi_speed | Maximum speed in km/h on a route_type `taxi` before triggering an `ExcessiveSpeed` warning |
+| other_speed | Maximum speed in km/h on a route_type `other` before triggering an `ExcessiveSpeed` warning |
```
Comment (thbar) [inline readme.md]:
```suggestion
| max_tramway_speed | Maximum speed in km/h on a route_type `tramway` before triggering an `ExcessiveSpeed` warning |
| max_subway_speed | Maximum speed in km/h on a route_type `subway` before triggering an `ExcessiveSpeed` warning |
| max_rail_speed | Maximum speed in km/h on a route_type `rail` before triggering an `ExcessiveSpeed` warning |
| max_bus_speed | Maximum speed in km/h on a route_type `bus` before triggering an `ExcessiveSpeed` warning |
| max_ferry_speed | Maximum speed in km/h on a route_type `ferry` before triggering an `ExcessiveSpeed` warning |
| max_cable_car_speed | Maximum speed in km/h on a route_type `cable car` before triggering an `ExcessiveSpeed` warning |
| max_gondola_speed | Maximum speed in km/h on a route_type `gondola` before triggering an `ExcessiveSpeed` warning |
| max_funicular_speed | Maximum speed in km/h on a route_type `funicular` before triggering an `ExcessiveSpeed` warning |
| max_coach_speed | Maximum speed in km/h on a route_type `coach` before triggering an `ExcessiveSpeed` warning |
| max_air_speed | Maximum speed in km/h on a route_type `air` before triggering an `ExcessiveSpeed` warning |
| max_taxi_speed | Maximum speed in km/h on a route_type `taxi` before triggering an `ExcessiveSpeed` warning |
| max_other_speed | Maximum speed in km/h on a route_type `other` before triggering an `ExcessiveSpeed` warning |
```

Comment (thbar) [reply readme.md]: Being more specific here allows us to add more checks later (just an idea)

File: readme.md [hunk 4]
```diff
@@ -292,6 +292,35 @@ You can then ask for a validation:
...
+
+### Example
+
+```yaml
+bus_speed: 120
+gondola_speed: 50
```
Comment (thbar) [inline readme.md]:
```suggestion
max_bus_speed: 120
max_gondola_speed: 50
```

File: src/bin/main.rs [hunk 5]
```diff
@@ -36,15 +36,22 @@ struct Opt {
         case_insensitive = true
     )]
     format: OutputFormat,
+    #[structopt(
+        short = "c",
+        long = "custom_rules",
```
Comment (thbar) [inline src/bin/main.rs]:
```suggestion
        long = "custom-rules",
```

File: src/custom_rules.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    pub rail_speed: Option<f64>,
+    pub bus_speed: Option<f64>,
+    pub ferry_speed: Option<f64>,
+    pub cable_car_speed: Option<f64>,
+    pub gondola_speed: Option<f64>,
+    pub funicular_speed: Option<f64>,
+    pub coach_speed: Option<f64>,
+    pub air_speed: Option<f64>,
+    pub taxi_speed: Option<f64>,
+    pub other_speed: Option<f64>,
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
pub struct CustomRules {
    pub max_tramway_speed: Option<f64>,
    pub max_subway_speed: Option<f64>,
    pub max_rail_speed: Option<f64>,
    pub max_bus_speed: Option<f64>,
    pub max_ferry_speed: Option<f64>,
    pub max_cable_car_speed: Option<f64>,
    pub max_gondola_speed: Option<f64>,
    pub max_funicular_speed: Option<f64>,
    pub max_coach_speed: Option<f64>,
    pub max_air_speed: Option<f64>,
    pub max_taxi_speed: Option<f64>,
    pub max_other_speed: Option<f64>,
```

File: src/custom_rules.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+pub fn custom_rules(file_path: Option<String>) -> CustomRules {
+    if let Some(path) = file_path {
+        let f = std::fs::File::open(path).expect("Could not open custom_rules file");
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
        let f = std::fs::File::open(path).expect("Could not open custom-rules file");
```

File: src/custom_rules.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        let d: CustomRules =
+            serde_yaml::from_reader(f).expect("custom_rules file is not a valid YAML file");
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
            serde_yaml::from_reader(f).expect("custom-rules file is not a valid YAML file");
```

File: src/custom_rules.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+}
+
+#[test]
+fn test_no_custom_rules() {
+    let file_path = None;
+    let custom_rules = custom_rules(file_path);
+    assert_eq!(None, custom_rules.bus_speed);
+    assert_eq!(None, custom_rules.air_speed);
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
    assert_eq!(None, custom_rules.max_bus_speed);
    assert_eq!(None, custom_rules.max_air_speed);
```

File: src/custom_rules.rs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+#[test]
+fn test_some_custom_rules() {
+    let file_path = Some(String::from("test_data/custom_rules/custom_rules.yml"));
+    let custom_rules = custom_rules(file_path);
+    assert_eq!(Some(10.), custom_rules.bus_speed);
+    assert_eq!(Some(100.5), custom_rules.gondola_speed);
+    assert_eq!(None, custom_rules.air_speed);
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
    assert_eq!(Some(10.), custom_rules.max_bus_speed);
    assert_eq!(Some(100.5), custom_rules.max_gondola_speed);
    assert_eq!(None, custom_rules.max_air_speed);
```

File: src/custom_rules.rs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+#[test]
+#[should_panic(expected = "Could not open custom_rules file")]
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
#[should_panic(expected = "Could not open custom-rules file")]
```

File: src/custom_rules.rs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+fn test_no_file() {
+    let file_path = Some(String::from("xxx"));
+    custom_rules(file_path);
+}
+
+#[test]
+#[should_panic(expected = "custom_rules file is not a valid YAML file")]
```
Comment (thbar) [inline src/custom_rules.rs]:
```suggestion
#[should_panic(expected = "custom-rules file is not a valid YAML file")]
```

File: src/validators/duration_distance.rs [hunk 13]
```diff
@@ -27,25 +28,31 @@ fn distance_and_duration(
...
+        Rail => custom_rules.rail_speed.unwrap_or(320.0),
+        Bus => custom_rules.bus_speed.unwrap_or(120.0),
+        Ferry => custom_rules.ferry_speed.unwrap_or(90.0), // https://en.wikipedia.org/wiki/List_of_HSC_ferry_routes
+        CableCar => custom_rules.cable_car_speed.unwrap_or(30.0),
+        Gondola => custom_rules.gondola_speed.unwrap_or(45.0), // https://fr.wikipedia.org/wiki/Vanoise_Express
+        Funicular => custom_rules.funicular_speed.unwrap_or(40.0),
+        Coach => custom_rules.coach_speed.unwrap_or(120.0),
+        Air => custom_rules.air_speed.unwrap_or(1_000.0),
+        Taxi => custom_rules.taxi_speed.unwrap_or(50.0),
+        Other(_) => custom_rules.other_speed.unwrap_or(120.0), // We suppose it’s a bus if it is invalid
```
Comment (thbar) [inline src/validators/duration_distance.rs]:
```suggestion
        Tramway => custom_rules.max_tramway_speed.unwrap_or(100.0),
        Subway => custom_rules.max_subway_speed.unwrap_or(140.0),
        Rail => custom_rules.max_rail_speed.unwrap_or(320.0),
        Bus => custom_rules.max_bus_speed.unwrap_or(120.0),
        Ferry => custom_rules.max_ferry_speed.unwrap_or(90.0), // https://en.wikipedia.org/wiki/List_of_HSC_ferry_routes
        CableCar => custom_rules.max_cable_car_speed.unwrap_or(30.0),
        Gondola => custom_rules.max_gondola_speed.unwrap_or(45.0), // https://fr.wikipedia.org/wiki/Vanoise_Express
        Funicular => custom_rules.max_funicular_speed.unwrap_or(40.0),
        Coach => custom_rules.max_coach_speed.unwrap_or(120.0),
        Air => custom_rules.max_air_speed.unwrap_or(1_000.0),
        Taxi => custom_rules.max_taxi_speed.unwrap_or(50.0),
        Other(_) => custom_rules.max_other_speed.unwrap_or(120.0), // We suppose it’s a bus if it is invalid
```

File: src/validators/duration_distance.rs [hunk 14]
```diff
@@ -193,12 +209,29 @@ fn test() {
     assert_eq!("near1", issues[4].object_id);
     assert_eq!(String::from("null"), issues[4].related_objects[0].id);
     assert_eq!(Some(String::from("Near1")), issues[4].object_name);
+
+    let custom_rules = custom_rules::CustomRules {
+        bus_speed: Some(1_000_000.0),
```
Comment (thbar) [inline src/validators/duration_distance.rs]:
```suggestion
        max_bus_speed: Some(1_000_000.0),
```

File: test_data/custom_rules/custom_rules.yml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+bus_speed: 10
+gondola_speed: 100.5
```
Comment (thbar) [inline test_data/custom_rules/custom_rules.yml]:
```suggestion
max_bus_speed: 10
max_gondola_speed: 100.5
```

**Review by thbar — CHANGES_REQUESTED** [review]
Bravo! I have made a few cosmetic suggestions, but otherwise looks good to me!

**Comment by fchabouis** [comment]
Thanks @thbar, I integrated your suggestions.

**Review by thbar — APPROVED** [review]

**Pull request merged**